ext/linux/osq_lock.h

```diff
@@ -0,0 +1,322 @@
+    atomic_set(&osq_ptr->tail, OSQ_UNLOCKED_VAL);
+    osq_nodepool_ptr = calloc(cores, sizeof(struct optimistic_spin_node));
+    if (osq_nodepool_ptr == NULL) exit(errno);
+}
+
+bool osq_lock(struct optimistic_spin_queue *lock);
+void osq_unlock(struct optimistic_spin_queue *lock);
+
+static inline bool osq_is_locked(struct optimistic_spin_queue *lock)
+{
+    return atomic_read(&lock->tail) != OSQ_UNLOCKED_VAL;
+}
+
+
+/*
+ * We use the value 0 to represent "no CPU", thus the encoded value
+ * will be the CPU number incremented by 1.
+ */
+static inline int encode_cpu(int cpu_nr)
+{
+    return cpu_nr + 1;
+}
+
+static inline int node_to_cpu(struct optimistic_spin_node *node)
+{
+    return node->cpu - 1;
+}
+
+static inline struct optimistic_spin_node * cpu_to_node(int encoded_cpu_val)
+{
+    int cpu_nr = encoded_cpu_val - 1;
+    return osq_nodepool_ptr + cpu_nr;
+}
+
+/*
+ * Get a stable @node->next pointer, either for unlock() or unqueue() purposes.
+ * Can return NULL in case we were the last queued and we updated @lock instead.
+ */
+static inline struct optimistic_spin_node *
+osq_wait_next(struct optimistic_spin_queue *lock,
+          struct optimistic_spin_node *node,
+          struct optimistic_spin_node *prev)
+{
+    /* sched_getcpu() return current cpu number */
+    int cpu_number = sched_getcpu();
+    if (cpu_number == -1) exit(errno);
+
+    struct optimistic_spin_node *next = NULL;
+    int curr = encode_cpu(cpu_number);
+    int old;
+
+    /*
+     * If there is a prev node in queue, then the 'old' value will be
+     * the prev node's CPU #, else it's set to OSQ_UNLOCKED_VAL since if
+     * we're currently last in queue, then the queue will then become empty.
+     */
+    old = prev ? prev->cpu : OSQ_UNLOCKED_VAL;
+
+    for (;;) {
+        if (atomic_read(&lock->tail) == curr &&
+            atomic_cmpxchg_acquire(&lock->tail, curr, old) == curr) {
+            /*
+             * We were the last queued, we moved @lock back. @prev
+             * will now observe @lock and will complete its
+             * unlock()/unqueue().
+             */
+            break;
+        }
+
+        /*
+         * We must xchg() the @node->next value, because if we were to
+         * leave it in, a concurrent unlock()/unqueue() from
+         * @node->next might complete Step-A and think its @prev is
+         * still valid.
+         *
+         * If the concurrent unlock()/unqueue() wins the race, we'll
+         * wait for either @lock to point to us, through its Step-B, or
+         * wait for a new @node->next from its Step-C.
+         */
+        if (node->next) {
+            next = xchg(&node->next, NULL);
+            if (next)
+                break;
+        }
+
+        cpu_relax();
+    }
+
+    return next;
+}
+
+bool osq_lock(struct optimistic_spin_queue *lock)
+{
+    /* sched_getcpu() return current cpu number */
+    int cpu_number = sched_getcpu();
```

Comment:
Same deal here, we should probably be using the user-space assigned threadnum.

Reply:
Modified.